### ISSUE TITLE
Why are the models different during pre training and finetune

### ISSUE THREAD

**Issue by SELECT-FROM** [bowang-lab/scGPT#103]
Hi,

Thank you for the amazing work! May I ask about why are the models different during pre training and finetune? I noticed that the strategy used during the finetune stage to document the pre-trained model was to load the same weights of the two models into the finetune model. May I ask if this is for any special purpose?

Thanks.

**Comment by subercui** [comment]
Hi, thank you for your question. Could you please clarify what you mean by "load the same weights of the two models"? Would you like to share a screenshot or point to the code lines related to your question?

**Comment by SELECT-FROM** [comment]
> Hi, thank you for your question. Could you please clarify what you mean by "load the same weights of the two models"? Would you like to share a screenshot or point to the code lines related to your question?

Thanks for your reply!  I'm sorry I didn't describe the problem clearly. The following code is what I saw in Tutorial Annotation.ipynb. My understanding is that it will take the intersection of the finetune model and the pre trained model architecture. Then it will extract the intersection part from the pre trained model weights and load it into the finetune model. To be honest, I really couldn't figure out the purpose of doing this. 
By the way, do you use raw counts during the pre training？

Thanks!

`if config.load_model is not None:
    try:
        model.load_state_dict(torch.load(model_file))
        logger.info(f"Loading all model params from {model_file}")
    except:
        # only load params that are in the model and match the size
        model_dict = model.state_dict()
        pretrained_dict = torch.load(model_file)
        pretrained_dict = {
            k: v
            for k, v in pretrained_dict.items()
            if k in model_dict and v.shape == model_dict[k].shape
        }
        for k, v in pretrained_dict.items():
            logger.info(f"Loading params {k} with shape {v.shape}")
        model_dict.update(pretrained_dict)
        model.load_state_dict(model_dict)`

**Comment by subercui** [comment]
Hi, this is all for technical reasons mostly related to pytorch. I'll try explain it:

You see the first option in the `try` is to simply load all the model params. The `model.load_state_dict` works quite strictly, in case there is any name or shape mismatch, it will throw an error. Now, in the annotation case, there are new layers added to the model. Remember we need to do the cell type classification, so a new classifier is appended to the model, which is a new component not in the pre-trained model. So for simplicity, we just implement the other branch to load all layers and params in the model that do match, and ignore the newly added classifier. You can check the messages printed out by the `logger.info`. All shared layers that make sense to be loaded should be successfully loaded this way, such as the parameters in the embedding layers and the transformer layers.

**Comment by SELECT-FROM** [comment]
Thanks for your reply! In the pretrain code,  the transformer_encoder is  FlashscGPTGenerator. But in the finetune stage the transformer_encoder is TransformerEncoder. The EncoderLayer in these two Transformer Encoders is not exactly the same. I want to know why it's designed like this. https://github.com/bowang-lab/scGPT/blob/4068d67caaac1e28d56964da68e0214817e38428/scgpt/model/model.py#L109-L133

**Comment by subercui** [comment]
Hi, in the pretraining stage, we implement our customized attention in `FlashscGPTGenerator`. You can find a hybrid implementation using pytorch and flash-attn at the https://github.com/bowang-lab/scGPT/blob/dev-temp/scgpt/model/flash_layers.py 

During fine-tuning, the training for current application tasks differs from the training in pretraining. Specifically, they do not need the customized masking in fine-tuning and can only work with the flash-attn part of the attention. This is actually much faster and memory-efficient. So we implemented and used the `FlashTransformerEncoderLayer`, https://github.com/bowang-lab/scGPT/blob/main/scgpt/model/model.py#L109 .

Notice that the two implementations share the same logic in most of the computations and share the naming of the parameters. So we can successfully load the pretraining weights during the fine-tuning
